enterprise/server/cmd/action_runner/main.go

```diff
@@ -0,0 +1,662 @@
+}
+func (invLog *invocationLog) Printf(format string, vals ...interface{}) {
+	invLog.Write([]byte(fmt.Sprintf(format+"\n", vals...)))
+}
+
+// buildEventPublisher publishes Bazel build events for a single build event stream.
+type buildEventPublisher struct {
+	streamID *bepb.StreamId
+	done     chan struct{}
+	events   chan *bespb.BuildEvent
+
+	mu  sync.Mutex
+	err error
+}
+
+func newBuildEventPublisher(streamID *bepb.StreamId) *buildEventPublisher {
+	return &buildEventPublisher{
+		streamID: streamID,
+		// We probably won't ever saturate this buffer since we only need to
+		// publish a few events for the actions themselves and progress events
+		// are rate-limited. Also, events are sent to the server with low
+		// latency compared to the rate limiting interval.
+		events: make(chan *bespb.BuildEvent, 256),
+		done:   make(chan struct{}, 1),
+	}
+}
+
+// Start the event publishing loop in the background. Stops handling new events
+// as soon as the first call to `Wait()` occurs.
+func (bep *buildEventPublisher) Start(ctx context.Context) {
+	go bep.run(ctx)
+}
+func (bep *buildEventPublisher) run(ctx context.Context) {
+	defer func() {
+		bep.done <- struct{}{}
+	}()
+
+	conn, err := dial(*besBackend)
+	if err != nil {
+		bep.setError(fmt.Errorf("error dialing bes_backend: %s", err))
+		return
+	}
+	defer conn.Close()
+	besClient := pepb.NewPublishBuildEventClient(conn)
+	stream, err := besClient.PublishBuildToolEventStream(ctx)
+	if err != nil {
+		bep.setError(fmt.Errorf("error opening build event stream: %s", err))
+		return
+	}
+
+	doneReceiving := make(chan struct{}, 1)
+	go func() {
+		for {
+			_, err := stream.Recv()
+			if err == nil {
+				continue
+			}
+			if err == io.EOF {
+				log.Println("Received all acks from server.")
+			} else {
+				log.Printf("Error receiving acks from the server: %s", err)
+				bep.setError(err)
+			}
+			doneReceiving <- struct{}{}
+			return
+		}
+	}()
+
+	for seqNo := int64(1); ; seqNo++ {
+		event := <-bep.events
+		if event == nil {
+			// Wait() was called, meaning no more events to publish.
+			// Send ComponentStreamFinished event before closing the stream.
+			start := time.Now()
+			err = stream.Send(&pepb.PublishBuildToolEventStreamRequest{
+				OrderedBuildEvent: &pepb.OrderedBuildEvent{
+					StreamId:       bep.streamID,
+					SequenceNumber: seqNo,
+					Event: &bepb.BuildEvent{
+						EventTime: ptypes.TimestampNow(),
+						Event: &bepb.BuildEvent_ComponentStreamFinished{ComponentStreamFinished: &bepb.BuildEvent_BuildComponentStreamFinished{
+							Type: bepb.BuildEvent_BuildComponentStreamFinished_FINISHED,
+						}},
+					},
+				},
+			})
+			log.Printf("BEP: published FINISHED event (#%d) in %s", seqNo, time.Since(start))
+
+			if err != nil {
+				bep.setError(err)
+				return
+			}
+			break
+		}
+
+		bazelEvent, err := ptypes.MarshalAny(event)
+		if err != nil {
+			bep.setError(fmt.Errorf("failed to marshal bazel event: %s", err))
```

Comment:
I'd like us to get away from fmt.Errorf where possible. Is there a status error that would make sense here?